57-monkfish/technical_specification.md

```diff
@@ -0,0 +1,76 @@
+# State Management Service (Monkfish)
+**Epic Type:** Implementation Epic
+
+Epic planning and implementation follow the
+[Epic Planning and Marathon SOP](https://docs.ghga-dev.de/main/sops/sop001_epic_planning.html).
+
+## Scope
+### Outline:
+The [Archive Test Bed](https://github.com/ghga-de/archive-test-bed) needs a way to
+consistently perform arbitrary CRUD operations
+in MongoDB that are outside the scope of the component services, but also unwieldy
+to manage in piecemeal code adaptations. Namely, the needs are:
+- setting a predefined state before running the tests (empty or prepopulate databases)
+- examining the state after or in between tests (whitebox testing)
+
+Right now, there's not a clean way to achieve either. 
+A dedicated service can offer a simple solution to these problems and allow test bed 
+processes to programmatically seed, reset, modify and examine databases through a single API.
+
+This service should **never** be deployed to production. It is *only* intended for use
+in the Testing and Staging environments, where there is no access to real data.  
+Despite this, there should be a way to restrict which databases and collections can be
+accessed with this service through configuration, and a simple API key (set in config)
+that can be used to authenticate requests.
+
+### Included/Required:
+The implementation of the SMS should include a token-secured RESTful API that interacts
+with MongoDB. There should be configuration to control access to databases and collections,
+as well as an API Key for authentication.
```

Comment:
You can already specify here the granularity of the access control config.

Suggestion: the setting should be a list of tuples containing (db-name, collection-name, permissions). permissions can we a string with the letters "crud". A value of none indicates no restriction. The first matching rule will be applied. A value of None can be used as a wildcard. If no rule matches, access will be denied.

E.g. `[('auth', 'users', 'c'), ('store', 'orders', 'ud'), (None, None, 'r')]` to allow creating users in db auth, updating or deleting orders in db store, and read from any other collection in any database. Or `[(None, None, None)]` to allow anything on any database.

Reply:
Or something similar, if you have a better idea.